fieldservice/models/fsm_order.py

```diff
@@ -306,52 +307,40 @@ def unlink(self):
             else:
                 raise ValidationError(_("You cannot delete this order."))
 
-    def _calc_scheduled_dates(self, vals):
+    @api.depends("scheduled_date_start", "scheduled_duration")
+    def _compute_scheduled_date_end(self):
         """Calculate scheduled dates and duration"""
-
-        if (
-            vals.get("scheduled_duration")
-            or vals.get("scheduled_date_start")
-            or vals.get("scheduled_date_end")
-        ):
-
-            if vals.get("scheduled_date_start") and vals.get("scheduled_date_end"):
-                new_date_start = fields.Datetime.from_string(
-                    vals.get("scheduled_date_start", False)
-                )
-                new_date_end = fields.Datetime.from_string(
-                    vals.get("scheduled_date_end", False)
-                )
-                hours = new_date_end.replace(second=0) - new_date_start.replace(
-                    second=0
-                )
-                hrs = hours.total_seconds() / 3600
-                vals["scheduled_duration"] = float(hrs)
-
-            elif vals.get("scheduled_date_end"):
-                hrs = (
-                    vals.get("scheduled_duration", False)
-                    or self.scheduled_duration
-                    or 0
-                )
-                date_to_with_delta = fields.Datetime.from_string(
-                    vals.get("scheduled_date_end", False)
-                ) - timedelta(hours=hrs)
-                vals["scheduled_date_start"] = str(date_to_with_delta)
-
-            elif vals.get("scheduled_duration", False) or (
-                vals.get("scheduled_date_start", False)
-                and (
-                    self.scheduled_date_start != vals.get("scheduled_date_start", False)
-                )
+        for order in self:
+            if (
+                order.scheduled_duration
+                or order.scheduled_date_start
+                or order.scheduled_date_end
```

Comment:
Something like this will be enough.
```suggestion
       
        for order in self:
            if order.scheduled_duration and order.scheduled_date_start:
                hours = order.scheduled_duration
                order.scheduled_date_end = order.scheduled_date_start + timedelta(
                        hours=hours
                    )
            else:
                order.scheduled_date_end = False;
```